### PR TITLE
gossiper: remove_endpoint: call on_dead notifications is endpoint was alive

### LINKED ISSUE #15178: gossiper: on_dead notifications are not called after remove_endpoint
Since 75d1dd3a762638686142e93d4a5cf2f6bc402eba
gossiper::convict will no longer call `mark_dead`
(e.g. when called from the failure detection loop
after a node is stopped following decommission)
and therefore the on_dead notification won't get called.

These are imporant to clean up after the node is dead
e.g. in storage_proxy::on_down which cancels all
respective write handlers.

This is visible since https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-release/339/testReport/bootstrap_test/TestBootstrap/Run_Dtest_Parallel_Cloud_Machines___FullDtest___full_split013___test_decommissioned_wiped_node_can_join/
```
        # Decommision the new node and wipe its data
        logger.info("Decommissioning node4")
        mark = node1.mark_log()
        node4.decommission()
        logger.debug("Stopping node4")
        node4.stop(wait_other_notice=False)
>       node1.watch_log_for("{} is now (dead|DOWN)".format(node4.address()), from_mark=mark)
...
>                       raise TimeoutError(time.strftime("%d %b %Y %H:%M:%S", time.gmtime()) + " [" + self.name + "] Missing: " + str(
                            [e.pattern for e in tofind]) + ":\n" + reads[:50] + f".....\nSee {filename} for remainder")
E                       ccmlib.node.TimeoutError: 24 Aug 2023 13:39:02 [node1] Missing: ['127.0.33.4 is now (dead|DOWN)']:
```

### PR DESCRIPTION
**Pull request by bhalevy**
Since 75d1dd3a762638686142e93d4a5cf2f6bc402eba
gossiper::convict will no longer call `mark_dead`
(e.g. when called from the failure detection loop
after a node is stopped following decommission)
and therefore the on_dead notification won't get called.

To make that explicit, if the node was alive before remove_endpoint erased it from _live_endpoint,
and it has an endpoint_state, call the on_dead notifications. These are imporant to clean up after the node is dead e.g. in storage_proxy::on_down which cancels all
respective write handlers.

This preferred over going through `mark_dead` as the latter marks the endpoint as unreachable, which is wrong in this case as the node left the cluster.

Fixes #15178

### CODE REVIEW

File: gms/gossiper.cc [hunk 1]
```diff
@@ -681,14 +681,28 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
         logger.info("removed {} from _seeds, updated _seeds list = {}", endpoint, _seeds);
     }
 
-    co_await mutate_live_and_unreachable_endpoints([endpoint] (gossiper& g) {
-        g._live_endpoints.erase(endpoint);
+    bool was_alive;
+    co_await mutate_live_and_unreachable_endpoints([endpoint, &was_alive] (gossiper& g) {
+        was_alive = g._live_endpoints.erase(endpoint);
         g._unreachable_endpoints.erase(endpoint);
```
Comment (avikivity) [inline gms/gossiper.cc:687]: Can remove this line now, no?

Comment (avikivity) [reply gms/gossiper.cc:687]: Sorry no.

File: gms/gossiper.cc [hunk 2]
```diff
@@ -681,14 +681,28 @@ future<> gossiper::remove_endpoint(inet_address endpoint, permit_id pid) {
...
     _ack_handlers.erase(endpoint);
     quarantine_endpoint(endpoint);
-    logger.debug("removing endpoint {}", endpoint);
+    logger.info("Removed endpoint {}", endpoint);
+
+    if (was_alive) {
+        try {
+            auto state = get_endpoint_state(endpoint);
+            logger.info("InetAddress {} is now DOWN, status = {}", endpoint, get_gossip_status(state));
+            co_await _subscribers.for_each([endpoint, state, pid] (shared_ptr<i_endpoint_state_change_subscriber> subscriber) -> future<> {
```
Comment (asias) [inline gms/gossiper.cc:698]: It is nicer to have a helper so we can share code with

```
future<> gossiper::mark_dead(inet_address addr, endpoint_state& local_state, permit_id pid) {
    logger.trace("marking as down {}", addr);
    verify_permit(addr, pid);
    endpoint_state state = local_state;
    co_await mutate_live_and_unreachable_endpoints([addr] (gossiper& g) {
        g._live_endpoints.erase(addr);
        g._unreachable_endpoints[addr] = now();
    });
    logger.info("InetAddress {} is now DOWN, status = {}", addr, get_gossip_status(state));
    co_await _subscribers.for_each([addr, state, pid] (shared_ptr<i_endpoint_state_change_subscriber> subscriber) -> future<> {
        co_await subscriber->on_dead(addr, state, pid);
        logger.trace("Notified {}", fmt::ptr(subscriber.get()));
    });
}
```

Comment (bhalevy) [reply gms/gossiper.cc:698]: fair enough.  Will add

**Pull request closed without merging**